btcutil: Add library functions for BIP-0352 silent payment _send_ support

guggero opened the pull request:
Adds library/API functionality for Silent Payment **send** support.

Implements the following BIPs (or pull requests to BIPs):
 - [BIP-0352 - Silent Payments](https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki) (added in https://github.com/bitcoin/bips/pull/1458)
 - [BIP-0374 - Discrete Log Equality Proofs](https://github.com/bitcoin/bips/blob/master/bip-0374.mediawiki) (added in https://github.com/bitcoin/bips/pull/1689)
 - [BIP-0375 - Sending Silent Payments with PSBTs](https://github.com/bitcoin/bips/blob/master/bip-0375.mediawiki) (added in https://github.com/bitcoin/bips/pull/1687)

Goal is to address **send** support first as that's rather straightforward. Receive support is much more involved as it requires scanning the chain in a specific manner.

guggero committed:
silentpayments: add test vectors, implement addr struct

guggero committed:
silentpayments: add send output support

guggero committed:
silentpayments: implement DLEQ proofs

guggero committed:
psbt: add global/input silent payment shares

guggero committed:
psbt: add global/input silent payment DLEQ proofs

guggero committed:
psbt: add output silent payment info

benma commented:
This is great work and very welcome. However, BIP-375 specifies exclusion for PSBTv0, so this will not be compatible with other PSBT libraries/wallets. Imho it would be best to focus on adding support for PSBTv2 first (https://github.com/btcsuite/btcd/issues/2328).

yyforyongyu commented on silentpayments/dleq.go:
```diff
@@ -0,0 +1,206 @@
+	B := ScalarMult(privateKey.Key, scanKey)
+	G := Generator()
+
+	proof, err := DLEQProof(privateKey, B, G, r, nil)
```
`CreateShare` passes the computed ECDH share (`a·scanKey`) as the `B` argument to `DLEQProof`. `DLEQProof` then computes `C = a·B`, so this proof is for `a²·scanKey`, not for the returned share. A verifier using `(A=privateKey.PubKey(), B=scanKey, C=share)` will reject it. This should pass `scanKey` to `DLEQProof` and return the separately computed share.

guggero replied:
Great catch! Fixed and added unit test.

yyforyongyu commented on silentpayments/dleq.go:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	// Let rand = hash_BIP0374/nonce(t || cbytes(A) || cbytes(C)).
+	rand := chainhash.TaggedHash(
```
Current BIP374 includes the optional message in the nonce derivation: `hash_BIP0374/nonce(t || cbytes(A) || cbytes(C) || m')`. This implementation omits `m`, which can leak `a` if the same aux randomness is reused with different messages.

guggero replied:
Yes, this was updated later on in the BIP. Fixed and did a full review on the latest version of the BIP.

yyforyongyu commented on silentpayments/input.go:
```diff
@@ -0,0 +1,195 @@
+		// These types are supported in any case.
+		return true, script, nil
+
+	case txscript.ScriptHashTy:
```
This accepts every P2SH output as silent-payment compatible. BIP352 only includes P2SH-P2WPKH in shared-secret derivation; arbitrary P2SH/multisig inputs should be excluded or otherwise identified precisely, otherwise sender and receiver can derive different secrets.

guggero replied:
Correct. Passing in the public key now so we can at least assert the address wasn't _made_ from an incorrect (e.g. uncompressed key).

yyforyongyu commented on silentpayments/address.go:
```diff
@@ -0,0 +1,337 @@
+		return nil, errors.New("invalid bech32 version")
+	}
+
+	regrouped, err := bech32.ConvertBits(data[1:], 5, 8, false)
```
`DecodeAddress` slices `data[1:]` and later reads `data[0]` before checking that `data` is non-empty. A checksum-valid Bech32m string with an empty data part can panic here instead of returning an error.

guggero replied:
Yes, good catch. Added a fuzz test as well and ran it for a few hours on a beefy machine.

yyforyongyu commented on psbt/partial_output.go:
```diff
@@ -144,6 +147,31 @@ func (po *POutput) deserialize(r io.Reader) error {
 				po.TaprootBip32Derivation, taprootDerivation,
 			)
 
+		case SilentPaymentV0InfoOutputType:
```
BIP375 defines `PSBT_OUT_SP_V0_INFO` with no key data, but this case does not reject non-empty `keyData`. Malformed PSBTs with extra key data are accepted and normalized instead of returning `ErrInvalidKeyData`.

guggero replied:
Yes, fixed.

yyforyongyu commented on psbt/partial_output.go:
```diff
@@ -144,6 +147,31 @@ func (po *POutput) deserialize(r io.Reader) error {
+
+			po.SilentPaymentInfo = info
+
+		case SilentPaymentV0LabelOutputType:
```
Same issue for `PSBT_OUT_SP_V0_LABEL`: the field has no key data per BIP375, so this should reject non-empty `keyData` before parsing the value.

yyforyongyu commented on silentpayments/dleq.go:
```diff
@@ -0,0 +1,206 @@
+func DLEQVerify(A, B, C, G *btcec.PublicKey, proof [64]byte, m *[32]byte) bool {
+	// Fail if any of is_infinite(A), is_infinite(B), is_infinite(C),
+	// is_infinite(G).
+	if !A.IsOnCurve() || !B.IsOnCurve() || !C.IsOnCurve() {
```
The comment quotes BIP374's requirement to fail if any of `A`, `B`, `C`, or `G` is infinite/off-curve, but the code only checks `A`, `B`, and `C`. Since `G` is caller-supplied, it should be validated too.

guggero replied:
Makes sense, fixed.

yyforyongyu commented on silentpayments/address.go:
```diff
@@ -0,0 +1,337 @@
+
+	idx := 0
+	grouped := make([][]Address, len(groups))
+	for _, group := range groups {
```
`GroupByScanKey` builds the returned slice by ranging over a map, so group order is randomized. Since `AddressOutputKeys` documents that recipients are ordered by output index and returns a flat result slice, this can make the returned output order nondeterministic for multiple scan-key groups.

guggero replied:
Very good point, addressed and unit tested.

yyforyongyu reviewed (commented):
Fired some agents to review it while catching up with the bip🤓

ping-localhost commented:
I currently use `btcutil.DecodeAddress` to validate if a provided Bitcoin address is valid. Seeing how it hasn't been changed, I'll have to make a fallback for Silent Payments once this is merged?

Is it maybe a good idea to add a more generic `ValidateAddress`-function that can deal with all types? Just brain dumping, so feel free to ignore me.

guggero committed:
psbt: add test vectors from BIP-375

guggero committed:
psbt: fix independent typo allowing duplicate xPubs

Fixes an independent issue discovered while reviewing and testing
the code that failed to detect duplicate xPubs because the wrong
variable (keyData holding the nil value of the global unsigned TX,
instead of the lowercase keydata that's the loop variable).

guggero commented:
> I currently use `btcutil.DecodeAddress` to validate if a provided Bitcoin address is valid. Seeing how it hasn't been changed, I'll have to make a fallback for Silent Payments once this is merged?
> 
> Is it maybe a good idea to add a more generic `ValidateAddress`-function that can deal with all types? Just brain dumping, so feel free to ignore me.

Hmm, that's a valid point. But I'm not sure what would be best here. Adding anything to the existing `DecodeAddress` in the now-extracted `address/v2` package can't be done, as that would introduce a circular package dependency.
And adding a new `DecodeAnyAddress` would need to return an `address.Address` interface. But that interface requires the `ScriptAddress() []byte` method to be defined that can be used to produce the `pkScript` of a receiver output.
But that's exactly the complication with Silent Payment addresses: They are dependent on the inputs being spent.
So for safety reasons, the `ScriptAddress` method would need to panic for Silent Payment addresses and any caller would need to know how to deal with them.

So perhaps a simple `silentpayments.IsValidAddress()` would be enough to make the detection easier?

guggero commented:
Thanks a lot for the review, @yyforyongyu! I addressed all your points and did a full (agentic) review against the three updated BIPs and made sure all updated test vectors also pass. Should be quite solid now.

ping-localhost commented:
> So perhaps a simple silentpayments.IsValidAddress() would be enough to make the detection easier?

Yeah, that would help already! Anything to make the process easier is very much appreciated.

guggero commented:
> > So perhaps a simple silentpayments.IsValidAddress() would be enough to make the detection easier?
> 
> Yeah, that would help already! Anything to make the process easier is very much appreciated.

Okay, I've added `IsValidAddress(string) bool`.
